server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -608,7 +608,11 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
     public void relocated(final Consumer<ReplicationTracker.PrimaryContext> consumer)
                                             throws IllegalIndexShardStateException, InterruptedException {
         assert shardRouting.primary() : "only primaries can be marked as relocated: " + shardRouting;
+        refreshListeners.disallowAdd();
         try {
+            if (refreshListeners.refreshNeeded()) {
+                refresh("relocated");
```

Comment:
As we always want to do the refresh after calling `disallowAdd`, I wonder if we should combine both into one method

Reply:
I couldn't find a neat way of doing that since we have the async case and the blocking case of acquiring all the permits and we want to enforce some `try-finally` semantics for allowing the listeners again for both now. I'm not sure it actually makes things more readable if we hide the handling of exceptions from `refresh(...)` in some other method. I can try finding a nice way though :)